### ISSUE TITLE
CUDA: binary classification probability saturation (prob→1.0) — non-deterministic bug in col-wise histogram kernel

### ISSUE THREAD

**Issue by adorkin87** [lightgbm-org/LightGBM#7231]
## Environment

| | |
|---|---|
| **LightGBM** | 4.6.0.99 (CUDA build) |
| **GPU** | NVIDIA GeForce RTX 5070 (12 GB, Blackwell) |
| **CUDA** | 12.9 |
| **Driver** | 580.126.09 |
| **NumPy** | 2.4.2 |
| **OS** | Ubuntu 25.10 (kernel 6.17.0-22-generic) |
| **Python** | 3.13.7 |

---

## Bug description

Binary classification with `device="cuda"` produces **saturated predictions** (`prob_max=1.0000`, hundreds of samples stuck at prob≈1.0) while CPU training on identical data/params gives healthy predictions (`prob_max<0.65`). AUC remains plausible (~0.80), which **masks the bug** in standard evaluation.

The bug is **non-deterministic**: ~3 out of 5 seeds trigger saturation. The same seed can give HEALTHY on one run and DEGENERATE on another.

**DEGENERATE** = `prob_max > 0.98` OR `n_samples_with_prob ≥ 1.0 > 10`  
**HEALTHY** = `prob_max < 0.80` (expected for this task)

---

### Trigger conditions (confirmed via 30-run diagnostic)

| Config | Status | prob_max | sat@1.0 |
|---|---|---|---|
| `device="cpu"` | HEALTHY | 0.60 | 0 |
| `device="cuda"` (default) | **~60% DEGENERATE** | 1.0000 | 136–1273 |
| `device="cuda", force_col_wise=True` | **DEGENERATE** | 1.0000 | 514 |
| `device="cuda", force_row_wise=True` | HEALTHY (usually) | 0.53 | 0 |
| `device="cuda", min_data_in_bin=30` | HEALTHY | 0.53 | 0 |
| `device="cuda", min_data_in_bin=3` (default) | **DEGENERATE** | 1.0000 | 24 |
| `device="cuda", min_data_in_bin=300` | **DEGENERATE** | 1.0000 | 68 |
| `device="cuda", max_bin=127` | HEALTHY (same run luck) | 0.53 | 0 |
| `device="cuda", gpu_use_dp=True` | HEALTHY (same run luck) | 0.52 | 0 |

**Key finding**: `force_col_wise=True` is the most stable trigger. LightGBM auto-selects col-wise for large feature counts (380 features), explaining why the default also triggers it intermittently.

### Seed sensitivity (device="cuda", 380 features, 1M–2M rows)

| Seed | Status | prob_max | sat@1.0 |
|---|---|---|---|
| 42 | **DEGENERATE** | 1.0000 | 1273 |
| 1 | **DEGENERATE** | 1.0000 | 646 |
| 7 | HEALTHY | 0.53 | 0 |
| 123 | **DEGENERATE** | 1.0000 | 640 |
| 999 | HEALTHY | 0.53 | 0 |

Same seed can produce different status across runs → **race condition** in CUDA reduction.

### Hyperparameters used (champion model params, but bug is not param-specific)

```python
params = {
    "objective":         "binary",
    "metric":            "auc",
    "device":            "cuda",
    "num_threads":       14,
    "seed":              42,
    "learning_rate":     0.009518840502449276,
    "num_leaves":        186,
    "max_depth":         7,
    "min_child_samples": 100,
    "feature_fraction":  0.5934692247305747,
    "bagging_fraction":  0.5179519471812114,
    "bagging_freq":      4,
    "lambda_l1":         0.5924602885530806,
    "lambda_l2":         0.1274528927681517,
    "min_gain_to_split": 0.001,
}
lgb.train(params, dtrain, num_boost_round=200)
```

### Code to reproduce (requires dataset, see below)

```python
import lightgbm as lgb
import numpy as np
import pandas as pd

df = pd.read_parquet("mre_minimal_1000k.parquet")
train = df[df["split"] == "train"]
val   = df[df["split"] == "val"]
feat_cols = [c for c in df.columns if c.startswith("f")]

X_tr, y_tr = train[feat_cols].to_numpy(dtype=np.float32), train["target"].to_numpy()
X_va, y_va = val[feat_cols].to_numpy(dtype=np.float32),   val["target"].to_numpy()

params = {
    "objective": "binary", "metric": "auc", "verbose": -1,
    "device": "cuda", "num_threads": 14, "seed": 42,
    "learning_rate": 0.009518840502449276, "num_leaves": 186, "max_depth": 7,
    "min_child_samples": 100, "feature_fraction": 0.5934692247305747,
    "bagging_fraction": 0.5179519471812114, "bagging_freq": 4,
    "lambda_l1": 0.5924602885530806, "lambda_l2": 0.1274528927681517,
    "min_gain_to_split": 0.001,
}

bst = lgb.train(params, lgb.Dataset(X_tr, label=y_tr), num_boost_round=200)
prob = bst.predict(X_va)

print(f"prob_max  = {prob.max():.6f}")          # Expected HEALTHY: < 0.70
print(f"sat@1.0   = {(prob >= 0.999999).sum()}") # Expected HEALTHY: 0
# DEGENERATE output: prob_max=1.000000, sat@1.0=646–1273
```

---

## Dataset

**1M rows × 380 features** (float32, binary target, pos_rate≈0.09).  
Derived from public Binance BTC/USDT 30-min OHLCV (2023-02 – 2024-12).  
Feature columns renamed `f0..f379` (no domain identifiers).  
File: `mre_minimal_1000k.parquet` — **743 MB** (too large for attachment).

👉 **Available on request** — please comment and I will share via Google Drive.

---

## What was ruled out

- **Synthetic data** (i.i.d. Gaussian, heavy-tail t(df=2), factor model, AR(1), mixed) — **does NOT reproduce** the bug across all tested sizes (200K–2M rows, 380 features). The bug requires real financial time-series data with temporal autocorrelation and specific gradient distributions.
- **Row shuffling** — destroys the bug. Temporal order of rows is required for reproduction, suggesting contiguous memory access patterns interact with the CUDA histogram kernel.
- **max_bin** — varying 31/63/127/255 does not reliably fix or trigger the bug.
- **gpu_use_dp=True** — does not fix the bug.
- **Feature pre-filter** — `feature_pre_filter=False` is NOT set; this is a different known issue.

---

## Root cause hypothesis

The **col-wise histogram CUDA kernel** has a race condition in its parallel reduction of gradient/hessian sums. When many temporally-autocorrelated rows map to the same histogram bin, concurrent thread blocks accumulate incorrect (near-zero) hessian sums. This causes leaf values `= -Σg / (Σh + λ)` to blow up, pushing raw scores to ±∞ within the first few trees. Subsequent iterations cannot recover because hessians for saturated samples ≈ 0.

Evidence:
1. `force_col_wise=True` reliably triggers; `force_row_wise=True` mostly safe
2. Same seed gives HEALTHY/DEGENERATE on different runs (non-deterministic)
3. Row shuffling prevents the bug (destroys temporal bin concentration)
4. AUC stays ~0.80 (saturation doesn't hurt ranking, only threshold-based selection)

---

## Workaround

Add a health-check after training and retry with a different seed:

```python
prob = bst.predict(X_val)
if prob.max() > 0.98 or (prob >= 0.999999).sum() > 10:
    raise RuntimeError("CUDA saturation detected — retrain with different seed")
```

Or force CPU for binary tasks on this hardware:
```python
params["device"] = "cpu"  # safe, but ~3× slower
```

**Comment by stiv-yakovenko** [comment]
I confirm the bug and here is a fix:

## Summary of `lightgbm-cuda-nan-fix.patch`

### Target Git revision

Repository: https://github.com/lightgbm-org/LightGBM.git

This patch was generated against and verified to apply to this exact commit:
`44818daddd0cd27c455a7974498d7af12635827a`.

LightGBM version at that commit: `4.7.0.99`.
Compatibility with other Git revisions has not been verified.

From a clean checkout at the commit above, apply the standalone patch with:

```bash
git apply --check lightgbm-cuda-nan-fix.patch
git apply lightgbm-cuda-nan-fix.patch
```

The full unified diff is included at the end of this document.

### Problem

In `src/io/cuda/cuda_tree.cu`, `AddPredictionToScoreKernel()` converts a bin
index from column coordinates to feature-local coordinates:

```cpp
bin = bin - min_bin + offset;
```

However, the NaN check compared the converted `bin` with the original
`max_bin`. These values could use different coordinate systems, causing
NaN values to go unrecognized and take the right branch instead of the
left branch specified by the model.

### Impact

Internal CUDA validation predictions differed from `model.predict()`.
DART uses the same tree traversal when removing and reapplying the
contributions of existing trees. The error therefore also corrupted
training scores and subsequent gradients, degrading model quality.

### Fix

Before checking for NaN, convert the maximum bin index to the same
coordinate system as `bin`:

```cpp
const uint32_t feature_max_bin = max_bin - min_bin + offset;
```

Before:

```cpp
missing_type == 2 && bin == max_bin
```

After:

```cpp
missing_type == 2 && bin == feature_max_bin
```

NaN values are now recognized correctly and routed according to
`default_left`. This fix requires no additional CUDA stream synchronization.
The synchronization proposed in `problem.txt` was tested separately and
did not resolve the observed failure on the supplied CSV.

### Validation

The patch includes `test_missing_value_prediction_consistency`: eight
cases (GBDT/DART and four value distributions) comparing internal
validation predictions with predictions from the trained model.
All eight cases failed on CUDA before the fix and pass after it.
Including related tests: 14/14 pass on CUDA and 14/14 pass on CPU.

The original `lightbm_bug.csv` was tested without modifying the file:
1,000 DART iterations, `seed=42`, the first `StratifiedGroupKFold` fold
grouped by user ID, the project's existing feature encoding, and
`float64` input:

```text
Model AUC:                        0.913293 -> 0.923604
Model log loss:                   0.418345 -> 0.343928
Maximum absolute difference
between internal predictions
and model.predict():              0.456198 -> 7.63e-15
```

These results cover one validation fold, not a full five-fold CV run.
SHA-256 verification confirmed that the original CSV remained unchanged.

Base revision: `44818daddd0cd27c455a7974498d7af12635827a`.

Test environment: LightGBM 4.7.0.99, RTX 5090, CUDA Toolkit 13.2.78.

Detailed report and artifacts: `artifacts/issue-7231/REPORT.md`.

### Full patch

Verbatim contents of `lightgbm-cuda-nan-fix.patch`:

```diff
diff --git a/src/io/cuda/cuda_tree.cu b/src/io/cuda/cuda_tree.cu
index 8b8c5ee3..11479aeb 100644
--- a/src/io/cuda/cuda_tree.cu
+++ b/src/io/cuda/cuda_tree.cu
@@ -377,7 +377,10 @@ __global__ void AddPredictionToScoreKernel(
         const uint32_t threshold_in_bin = cuda_threshold_in_bin[node];
         const int8_t missing_type = GetMissingTypeCUDA(decision_type);
         const bool default_left = ((decision_type & kDefaultLeftMask) > 0);
-        if ((missing_type == 1 && bin == default_bin) || (missing_type == 2 && bin == max_bin)) {
+        // bin is feature-local after the conversion above, while max_bin is
+        // still in column coordinates. Convert the NaN bin before comparing.
+        const uint32_t feature_max_bin = max_bin - min_bin + offset;
+        if ((missing_type == 1 && bin == default_bin) || (missing_type == 2 && bin == feature_max_bin)) {
           if (default_left) {
             node = cuda_left_child[node];
           } else {
diff --git a/tests/python_package_test/test_engine.py b/tests/python_package_test/test_engine.py
index 2f80b21e..e6fd422d 100644
--- a/tests/python_package_test/test_engine.py
+++ b/tests/python_package_test/test_engine.py
@@ -238,6 +238,40 @@ def test_missing_value_handle_more_na():
     assert evals_result["valid_0"]["l2"][-1] == pytest.approx(ret)
 
 
+@pytest.mark.parametrize("boosting_type", ["gbdt", "dart"])
+@pytest.mark.parametrize("most_frequent_value", [-1.0, 0.0, 1.0, np.nan])
+def test_missing_value_prediction_consistency(boosting_type, most_frequent_value):
+    # NaNs go left, while the largest non-missing values go right. Vary the
+    # most frequent bin to exercise the column-to-feature bin offset.
+    values = np.array([-1.0, 0.0, 1.0, np.nan, most_frequent_value, most_frequent_value])
+    X = np.tile(values, 100).reshape(-1, 1)
+    y = (X[:, 0] >= 0).astype(np.float64)
+    train = lgb.Dataset(X, y)
+    valid = lgb.Dataset(X.copy(), y, reference=train)
+    cached_predictions = []
+
+    def capture_predictions(predictions, _):
+        cached_predictions.append(predictions.copy())
+        return "unused", 0.0, False
+
+    model = lgb.train(
+        {
+            "objective": "regression",
+            "metric": "None",
+            "boosting_type": boosting_type,
+            "num_leaves": 2,
+            "min_data_in_leaf": 1,
+            "skip_drop": 0.0,
+            "verbosity": -1,
+        },
+        train,
+        num_boost_round=10,
+        valid_sets=[valid],
+        feval=capture_predictions,
+    )
+    np.testing.assert_allclose(cached_predictions[-1], model.predict(X), rtol=1e-12, atol=1e-12)
+
+
 def test_missing_value_handle_na():
     x = [0, 1, 2, 3, 4, 5, 6, 7, np.nan]
     y = [1, 1, 1, 1, 0, 0, 0, 0, 1]
```
